### PR TITLE
fixes for enabling write barrier testing on windows

### PR DESCRIPTION
**Pull request by leirocks**
restore the compiler assert, thanks @curtisman  for the better solution
use VirtualQuery to get the reserve limit for the stack

### CODE REVIEW

**Comment by leirocks** [comment]
@dotnet-bot test Linux tests please

**Comment by leirocks** [comment]
@dotnet-bot test Linux tests please

**Comment by leirocks** [comment]
@dotnet-bot test this please

**Comment by leirocks** [comment]
@dotnet-bot test Linux tests please

**Comment by leirocks** [comment]
@dotnet-bot test Linux tests please

**Comment by leirocks** [comment]
@dotnet-bot test Linux tests please

**Comment by leirocks** [comment]
@dotnet-bot test Linux tests please

**Comment by leirocks** [comment]
@dotnet-bot test this please

**Comment by leirocks** [comment]
@dotnet-bot test OSX please

File: lib/Common/Memory/RecyclerWriteBarrierManager.cpp [hunk 1]
```diff
@@ -71,10 +74,27 @@ X64WriteBarrierCardTableManager::OnThreadInit()
...
+    // the AllocationBase from the VirtualQuery is the whole region possibly used for stack
+    // the BaseAddress from VirtualQuery is current commited base address, we can also query the region of GuardPages but seems not necessary
+
+    stackEnd = (char*)memInfo.AllocationBase;
+    Assert(memInfo.AllocationProtect == PAGE_READWRITE);
+#else
+    pthread_attr_t attr;
+    size_t stacksize;
+    pthread_attr_init(&attr);
+    pthread_attr_getstacksize(&attr, &stacksize); // assuming no one is calling pthread_attr_setstacksize afterwards
```
Comment (obastemur) [inline lib/Common/Memory/RecyclerWriteBarrierManager.cpp]: Besides we tend to not do this `ifdef _WIN32` stuff under the core. All the necessary `stack` related functionality is available under PAL or PlatformAgnostic

**Comment by obastemur** [comment]
It could be nice to have these commits squashed before merged though.

File: lib/Common/Memory/RecyclerWriteBarrierManager.cpp (hunk 1 above)
Comment (leirocks) [inline lib/Common/Memory/RecyclerWriteBarrierManager.cpp]: I'll try that, thanks

**Comment by leirocks** [comment]
@obastemur yes I'll squash it. there's lots of crashes on windows, which happened with combining last two PR - one from JC and one from me. and hard to repro on my work machine. trying the commits to verify the fix here.

**Comment by leirocks** [comment]
replaced by #2140, as I can't update this one on my another computer. and squashed the commits

**Pull request closed without merging**